src/internal/document_api/mod.rs

```diff
@@ -435,53 +500,177 @@ pub fn encrypt_document<
     grant_to_author: bool,
     user_grants: &'a Vec<UserId>,
     group_grants: &'a Vec<GroupId>,
-    policy_grant: Option<&PolicyGrant>,
+    policy_grant: Option<&'a PolicyGrant>,
 ) -> impl Future<Item = DocumentEncryptResult, Error = IronOxideErr> + 'a {
     let (dek, doc_sym_key) = transform::generate_new_doc_key(recrypt);
     let doc_id = document_id.unwrap_or(DocumentId::goo_id(rng));
+    aes::encrypt_future(rng, &plaintext.to_vec(), *doc_sym_key.bytes())
+        .join(resolve_keys_for_grants(
+            auth,
+            user_grants,
+            group_grants,
+            policy_grant,
+            grant_to_author,
+        ))
+        .and_then(move |(encrypted_doc, (grants, key_errs))| {
+            encrypt_document_core(
```

Comment:
Didn't `encrypt_future` actually encrypt the document?

Reply:
`encrypt_future` aes encrypted the data. `encrypt_document_core` recrypt encrypts the DEK

Reply:
`encrypt_document_core` as a name didn't really help me understand that. Seems almost counter-intuitive - `encrypt_document_core` encrypts the stuff that isn't the core of the document.

Reply:
Ah. It was meant to be the "core" of the encryption not the document. Renamed to `recrypt_document' to see if that makes more sense.